src/lib.rs

```diff
@@ -60,6 +88,166 @@ impl Bencher {
         }
         self.elapsed = start.elapsed();
     }
+
+    /// Times a `routine` that requires some `setup` on each iteration.
+    ///
+    /// For example, use this loop to benchmark sorting algorithms because they require unsorted
+    /// data on each iteration.
+    ///
+    /// # Example
+    ///
+    /// ``` no_run
+    /// extern crate criterion;
+    ///
+    /// use criterion::Bencher;
+    ///
+    /// fn create_data() -> Vec<u64> {
+    ///     # vec![]
+    ///     // ...
+    /// }
+    ///
+    /// // This should be a deterministic (i.e. not random) function
+    /// fn scramble(data: &mut [u64]) {
+    ///     // ...
+    /// }
+    ///
+    /// // The sorting algorithm to test
+    /// fn sort(data: &mut [u64]) {
+    ///     // ...
+    /// }
+    ///
+    /// fn benchmark(b: &mut Bencher) {
+    ///     let ref mut data = create_data();
+    ///
+    ///     b.iter_with_setup(|| scramble(data), |_| sort(data))
+    /// }
+    ///
+    /// # fn main() {}
+    /// ```
+    ///
+    /// # Timing loop
+    ///
+    /// ``` ignore
+    /// elapsed = 0;
+    /// for _ in 0..iters {
+    ///     let input = setup();
+    ///
+    ///     let start = Instant::now();
+    ///     let output = routine(input);
+    ///     let elapsed_in_iter = start.elapsed();
+    ///
+    ///     mem::drop(output);
+    ///
+    ///     elapsed += elapsed_in_iter;
+    /// }
+    /// ```
+    ///
+    /// # Timing model
+    ///
+    /// Note that `Bencher` also times the `Instant::now` function. Criterion will warn you (NOTE
+    /// not yet implemented) if the runtime of `routine` is small or comparable to the runtime of
+    /// `Instant::now` as this indicates that the measurement is useless.
+    ///
+    /// ``` text
+    /// elapsed = iters * (Instant::now + routine)
+    /// ```
+    pub fn iter_with_setup<I, O, S, R>(&mut self, mut setup: S, mut routine: R)
+        where S: FnMut() -> I,
+              R: FnMut(I) -> O
+    {
+        self.elapsed = Duration::from_secs(0);
+        for _ in 0..self.iters {
+            let input = setup();
+
+            let start = Instant::now();
+            let output = test::black_box(routine(input));
```

Comment:
Would it make sense to add `black_box` around the input as well? I'm not sure about this, just want to prevent rustc to optimize the call `routine(input)`